staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go

```diff
@@ -161,6 +161,18 @@ type JSONSchemaProps struct {
 	//      Atomic maps will be entirely replaced when updated.
 	// +optional
 	XMapType *string `json:"x-kubernetes-map-type,omitempty" protobuf:"bytes,43,opt,name=xKubernetesMapType"`
+
+	// x-kubernetes-validator describes a list of validation rules for expression validation.
+	XValidator []CELValidationRules `json:"x-kubernetes-validator,omitempty" protobuf:"bytes,44,rep,name=xKubernetesValidator"`
+}
+
+// CELValidationRules is used for expression validation.
+// ref: https://github.com/google/cel-spec
+type CELValidationRules struct {
+	// Rule represents the validation rule which will be evaluated by CEL.
+	Rule string `json:"rule,omitempty" protobuf:"bytes,1,opt,name=rule"`
+	// Message represents the message displayed when validation failed.
+	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
```

Comment:
The message returned to user would be the detailed error together with the ValidationRule.Message if it's set. I think it should be ok if user does not set message(they will still get error message)

Reply:
I think optional is appropriate. We expect a significant portion of the rules to be short and self descriptive. For these rules, I don't expect developers to give a message that is more readable than the rule itself, especially when they're only adding a message because the API requires it.

Reply:
As long as we include the rule as part of the CRD validation error, I suppose that's fine.